stingray/utils.py

```diff
@@ -1139,3 +1137,375 @@ def interpret_times(time, mjdref=0):
+
+    for t in range(tracks.shape[1]):
+        slicearr = np.array(
+            [
+                (tracks[dim, t] - ranges[dim, 0]) * delta[dim]
+                for dim in range(tracks.shape[0])
+            ]
+        )
+
+        good = np.all((slicearr < bins) & (slicearr >= 0))
+        slice_int[:] = slicearr
+
+        if good:
+            curr = index_arr(H, slice_int)
+            index_set_arr(H, slice_int, curr + 1)
+
+    return H
+
+
+def histnd_numba_seq(tracks, bins, ranges):
+    """
+    Examples
+    --------
+    >>> x = np.random.uniform(0., 1., 100)
+    >>> y = np.random.uniform(2., 3., 100)
+    >>> z = np.random.uniform(4., 5., 100)
+    >>> # 2d example
+    >>> H, _, _ = np.histogram2d(x, y, bins=np.array((5, 5)),
+    ...                          range=[(0., 1.), (2., 3.)])
+    >>> alldata = np.array([x, y])
+    >>> Hn = histnd_numba_seq(alldata, bins=np.array([5, 5]),
+    ...                       ranges=np.array([[0., 1.], [2., 3.]]))
+    >>> assert np.all(H == Hn)
+    >>> # 3d example
+    >>> H, _ = np.histogramdd((x, y, z), bins=np.array((5, 6, 7)),
+    ...                       range=[(0., 1.), (2., 3.), (4., 5)])
+    >>> alldata = np.array([x, y, z])
+    >>> Hn = hist3d_numba_seq(alldata, bins=np.array((5, 6, 7)),
+    ...                       ranges=np.array([[0., 1.], [2., 3.], [4., 5.]]))
+    >>> assert np.all(H == Hn)
+    """
+    H = np.zeros(tuple(bins), dtype=np.uint64)
+    slice_int = np.zeros(len(bins), dtype=np.uint64)
+
+    return _histnd_numba_seq(H, tracks, bins, ranges, slice_int)
+
+
+if HAS_NUMBA:
+
+    def histogram2d(*args, **kwargs):
+        if "range" in kwargs:
+            kwargs["ranges"] = kwargs.pop("range")
+        return hist2d_numba_seq(*args, **kwargs)
+
+    def histogram(*args, **kwargs):
+        if "range" in kwargs:
+            kwargs["ranges"] = kwargs.pop("range")
+        return hist1d_numba_seq(*args, **kwargs)
+
+
+else:
+
+    def histogram2d(*args, **kwargs):
+        return histogram2d_np(*args, **kwargs)[0]
+
+    def histogram(*args, **kwargs):
+        return histogram_np(*args, **kwargs)[0]
+
+
+def equal_count_energy_ranges(energies, n_ranges, emin=None, emax=None):
+    """Find energy ranges containing an approximately equal number of events.
+
+    Examples
+    --------
+    >>> energies = np.random.uniform(0, 10, 1000000)
+    >>> edges = equal_count_energy_ranges(energies, 5, emin=0, emax=10)
+    >>> np.allclose(edges, [0, 2, 4, 6, 8, 10], atol=0.05)
+    True
+    >>> edges = equal_count_energy_ranges(energies, 5)
+    >>> np.allclose(edges, [0, 2, 4, 6, 8, 10], atol=0.05)
+    True
+    >>> edges = equal_count_energy_ranges(energies, 0)
+    >>> np.allclose(edges, [0, 10], atol=0.05)
+    True
+    """
+    need_filtering = False
+    if emin is None:
+        emin = energies.min()
+        need_filtering = True
+    if emax is None:
+        emax = energies.max()
+        need_filtering = True
+    if need_filtering:
```

Comment:
I don't understand this logic. `need_filtering` is False by default. It will be set to true if one of `emin` or `emax` are `None`, in which case the filtering will happen between `emin` and `emax`. But if `emin` and `emax` are *not* set, then we don't *need* filtering, because they'll be automatically set to the smallest and largest energy value.

Conversely, if `emin` and `emax` are *both* set, there won't be any filtering at all? Because `need_filtering` would never be set to `True`. But that seems counter to what you're trying to do here?

Reply:
Yeah I don't know what I was thinking :D